# Java异常类

在程序运行中，任何中断正常流程的因素都被认为是异常。按照面对对象的思想，Java语言用Java类来描述异常。所有的异常类的祖先类为`java.lang.Throwable`类，它的实例表示具体的异常对象，可以通过`throw`语句抛出。

## 异常体系结构

Throwable
- Error：Java虚拟机无法解决的严重问题。如：JVM系统内部错误、资源耗尽等严重情况。比如：StackOverflowError和OOM。一般不编写针对性的代码进行处理。
- Exception：其它因编程错误或偶然的外在因素导致的一般性问题，可以使用针对性的代码进行处理。

例如
编译时异常（checked）是指编译器要求必须处置的异常。即程序在运行时由于外界因素造成的一般性异常。编译器要求Java程序必须捕获或声明所有编译时异常。对于这类异常，如果程序不处理，可能会带来意想不到的结果。
- IOException：IO异常
- FileNotFoundException：文件未找到异常
- ClassNotFoundExcption：类未找到

运行时异常（unchecked,RuntimeException）是指编译器不要求强制处置的异常。一般是指编程时的逻辑错误，是程序员应该积极避免其出现的异常。java.lang.RuntimeException类及它的子类都是运行时异常。对于这类异常，可以不作处理，因为这类异常很普遍，若全处理可能会对程序的可读性和运行效率产生影响。
-   NullpointerExceotion(空指针异常)
-   ArrayIndexOutfBoundsException（数组角标越界）
-   ClasssCastException（类型转换异常）
-   NumberForException（数值转换异常）
-   InputMismatchException
-   ArithmeticException

# 异常处理机制

**Java采用的异常处理机制，是将异常处理的程序代码集中在一起， 与正常的程序代码分开，使得程序简洁、优雅，并易于维护。**

Java提供的是异常处理的抓抛模型。
**抓抛模型**
过程一“抛”：程序在正常执行的过程中，一旦出现异常，就会在异常代码处生成一个对应异常类的对象，并将此对象抛出。一旦抛出对象以后，其后的代码就不再执行。

**关于异常对象的产生**
①：由虚拟机自动生成：程序运行过程中，虚拟机检测到程序发生了问题，如果在当前代码中没有找到相应的处理程序，就会在后台自动创建一个对应异常类的实例 对象并抛出自动抛出

②：由开发人员手动创建：`Exception exception = new ClassCastException();`
创建好的异常对象不抛出对程序没有任何影响，和创建一个普通对象一样

过程二“抓”：可以理解为异常的处理方式：①`try-catch-finally` ②`throws`

如果一个方法内抛出异常，该异常对象会被抛给调用者方法中处理。如果异常没有在调用者方法中处理，它继续被抛给这个调用方法的上层方法。这个过程将一直继续下去，直到异常被处理。 这一过程称为捕获(catch)异常。

如果一个异常回到main()方法，并且main()也不处理，则程序运行终止。

程序员通常只能处理Exception，而对Error无能为力。

## 异常处理方式之一： try-catch-finally的使用

```java
try {
	// 可能出现异常的代码
} catch (异常类型1 变量名1) {
	// 处理异常的方式1
} catch (异常类型2 变量名2) {
	// 处理异常的方式2
} catch (异常类型3 变量名3) {
// 处理异常的方式3
// .....
finally{
// 一定会执行的代码
}
```

try块： 用于捕获异常。其后可接零个或多个 catch 块，如果没有 catch 块，则必须跟一个 finally 块。

catch块： 用于处理 try 捕获到的异常。

finally块： 无论是否捕获或处理异常，finally块里的语句都会被执行。当在 try 块或 catch 块中遇到 return 语句时finally 语句块将在方法返回之前被执行。

说明：
1.  finally是可选的
2.  使用try将可能会出现异常代码包装起来，在执行过程中，一旦出现异常，就会生成一个对应异常类的对象，根据此对象的类型，去catch进行匹配
3.  一旦try中的异常对象匹配到某一个catch时，就进入catch进行异常的处理。一旦处理完成，就跳出当前的try-catch结构（在没有写finally的情况）。继续执行其后的代码

异常类型关系
- catch中的异常类型如果没有子父关系类型，则谁声明在上，谁声明在下无所谓
- catch中的异常类型如果满足子父关系类型，则要求子类一个声明在父类的上面。否则，报错

常用的异常对象处理的方式
- `String getMessage()` :返回类型的异常信息
- `printStackTrace()`：打印跟踪方法调用栈而获得的详细异常信息。在程序调试阶段，此方法可用于跟踪错误

在try结构中声明的变量，再出了try以后，就不能再被调用
try-catch-finally可以嵌套

**体会**
使用try-catch-finally处理编译时异常，使得程序在编译时就不再报错，但是运行仍可能报错。相当于我们使用try-catch-finally将一个编译时可能出现的异常，延迟到运行时出现

开发中，由于运行时异常比较常见，所以我们通常不针对运行时异常编写try-catch-finally了。针对于编译时异常，我们说一定要考虑异常的处理

**try-catch-finally中finally的使用**
-   finally是可选的
-   语句，catch中有return语句等情况
-   像数据库连接，输入输出流，网络编程Socket等资源,JVM是不能自动的回收的，我们需要自己手动的进行资源释放。此时的资源释放，就需要声明在finally中

在以下 3 种特殊情况下，finally 块不会被执行：
- 在 try 或 finally块中用了 `System.exit(int)`退出程序。但是，如果 System.exit(int) 在异常语句之后，finally 还是会被执行
- 程序所在的线程死亡。
- 关闭 CPU。

注意： 当 try 语句和 finally 语句中都有 return 语句时，在方法返回之前，finally 语句的内容将被执行，并且 finally 语句的返回值将会覆盖原始的返回值。


```java
public static int f(int value){
	try{
	        return value*value;
	}finally{
	        if(value==2){
	        return 0;
	        }
	  }
}  
```
如果调用 f(2)，返回值将是 0，因为 finally 语句的返回值覆盖了 try 语句块的返回值。

Throwable 类常用方法
- `public string getMessage()`:返回异常发生时的简要描述
- `public string toString()`:返回异常发生时的详细信息
- `public string getLocalizedMessage()`:返回异常对象的本地化信息。使用 Throwable 的子类覆盖这个方法，可以生成本地化信息。如果子类没有覆盖该方法，则该方法返回的信息与 getMessage（）返回的结果相同
- `public void printStackTrace()`:在控制台上打印 Throwable 对象封装的异常信息

## 异常处理方式之二：throws + 异常类型

“throws + 异常类型”写在方法的声明处。指明此方法执行时，可能会抛出的异常类型

如果一个方法(中的语句执行时)可能生成某种异常，会在异常代码处生成一个异常类的对象，**但是并不能确定如何处理这种异常，则此方法应显示地声明抛出异常，表明该方法将不对这些异常进行处理,而由该方法的调用者负责处理**,异常代码后续的代码，就不再执行

在方法声明中用throws语句可以声明抛出异常的列表，throws后面的异常类型可以是方法中产生的异常类型，也可以是它的父类。

**两种处理方式的区别**
- try-catch-finally是真正的将异常给处理掉了
- throws的方式只是将异常抛给了方法的调用者，并没有将异常处理掉。

**开发中如何选择使用try-catch-finally 还是使用throws？**
- 如果父类中被重写的方法没有throws方式处理异常，则子类重写的方法也不能throws，意味着如果子类重写的方法中有异常，必须使用try-catch-finally方式处理。
- 执行的方法a中，先后调用了另外的几个方法，这几个方法是递进关系执行的。我们建议这几个方法使用throws的方式进行处理。而执行的方法a可以考虑使用try-catch-finally方式进行处理。

**手动抛出异常对象**
在程序执行中，除了自动抛出异常对象的情况之外，我们还可以手动的throw一个异常类的对象

**throws和throw区别**
throw表示抛出一个异常类的对象，生成异常对象的过程。声明在方法体内
throws属于异常处理的一种方式，声明在方法的声明处

如何自定义异常类？
1.  继承于现有的异常结构：RuntimeExcption，Exception
2.  提供全局常量：serialVersionUID
3.  提供重载的构造器

例子
```java
public class MyExceotion extends RuntimeExcpetion{
        static final long serialVersionUID=-7034871901515766939L;

        public MyExceotion(){
        }

        public MyExceotion(String msg){
                super(msg);
        }
}
```